### PR TITLE
downloads: Creates new download site to curate details from the GitHub releases page for SEO purposes

### PR DESCRIPTION
**Pull request by wgordon17**

### CODE REVIEW

**Comment by wgordon17** [comment]
Closes #66

File: README.md [hunk 1]
```diff
@@ -17,6 +17,10 @@ should be able to preview your changes at http://localhost:4567
 
     bundle exec middleman server
 
+Dockerfile is included for processing PHP content. To run the PHP server execute the following from the root repo directory:
```
Comment (jiri-fiala) [inline README.md]: Having trouble with this, hitting 403s with apache running as described above. Overall, looks like an over-complicated thing to me, just for a local testing (as it needs stuff [like this](https://github.com/openshift/openshift.org/pull/68/files#diff-f828450012b9c91ffbf63af17fa0c66cR11)). Why not using something like [middleman-php](https://github.com/lord/middleman-php) for development?

File: source/github.php [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+} else {
+    $github_json = ['headers' => ['Last-Modified' => false]];
+}
+
+// Send API request to GitHub, including authentication information to ensure rate limits are higher
+// See https://developer.github.com/v3/#increasing-the-unauthenticated-rate-limit-for-oauth-applications for details
+$curl = curl_init();
+$url = 'https://api.github.com/repos/openshift/origin/releases/latest';
+$data = ['client_id' => '7e65d957b9bb631e53e6',
+         'client_secret' => '6ce97999e104ed75fded7328485a5f0e36357a46'];
```
Comment (jiri-fiala) [inline source/github.php]: This [should not be shared](https://developer.github.com/guides/basics-of-authentication/#registering-your-app) publicly in the repo.

File: source/js/download.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+    $.get(php_url, function(data) {
+        var jdata = JSON.parse(data),
+            linux_oc = $('#oc-platforms .linux-platform'),
+            mac_oc = $('#oc-platforms .mac-platform'),
+            windows_oc = $('#oc-platforms .windows-platform'),
+            linux_server = $('#server-platforms .linux-platform'),
+            asset, index, append_html;
+
+        console.log(jdata['assets']);
```
Comment (jiri-fiala) [inline source/js/download.js]: Should be removed from production version?

File: source/js/download.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        
+        for (var i = 0; i < jdata['assets'].length; i++) {
+            asset = jdata['assets'][i];
+            if ((index = asset['name'].indexOf('client-tools')) >= 0) {
+                append_html = '<a href="' + asset['browser_download_url'] + '">' + asset['name'].substring(index + 13) + '</a><br>';
```
Comment (jiri-fiala) [inline source/js/download.js]: `substring(index+13)` does not seem as something that will last working for ages; missing some comments there at least.

**Review by jiri-fiala — CHANGES_REQUESTED** [review]
The client secret hardcoded in the repo looks like a red flag to me. I will also need some assistance with the apache permissions.

File: 000-default.conf [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+ Header set Access-Control-Allow-Origin "*"
+ Header always set Access-Control-Allow-Methods "POST, PUT, GET, DELETE, OPTIONS"
+ Header always set Access-Control-Allow-Headers "Content-Type"
+
+ <Directory /var/www>
+ Options Indexes FollowSymLinks
+ AllowOverride All
+ Require all granted
+ </Directory>
+</VirtualHost>
```
Comment (jsvgoncalves) [inline 000-default.conf]: new line

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from nimmis/apache-php5:latest
+
+COPY ./000-default.conf /etc/apache2/sites-available/000-default.conf
+RUN a2enmod headers
+
+EXPOSE 80
+
+CMD ["/usr/sbin/apache2ctl", "-D", "FOREGROUND"]
```
Comment (jsvgoncalves) [inline Dockerfile]: new line

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,8 @@
+from nimmis/apache-php5:latest
+
+COPY ./000-default.conf /etc/apache2/sites-available/000-default.conf
```
Comment (jsvgoncalves) [inline Dockerfile]: If we will use docker, maybe it would be a good idea to move it to some other folder? e.g. `./docker/000-default.conf`

File: README.md (hunk 1 above)
Comment (jsvgoncalves) [reply README.md]: Agreed

File: source/css/custom.scss [hunk 8]
```diff
@@ -36,14 +36,68 @@ a.strong {
 	padding-left: 0;
 	float: left;
 	> li {
-		margin: 0 50px 0 0;
+		margin: 0 50px 15px 0;
```
Comment (jsvgoncalves) [inline source/css/custom.scss]: Is this relevant?

File: source/css/custom.scss [hunk 9]
```diff
@@ -36,14 +36,68 @@ a.strong {
...
 		> a {
 			font-size: 18px;
 			text-shadow: 1px 1px 1px rgba(0, 0, 0, .5);
 		}
 	}
 }
 
+.code {
```
Comment (jsvgoncalves) [inline source/css/custom.scss]: Please fix the indentation, it's a mix of tabs and spaces

File: source/css/custom.scss [hunk 10]
```diff
@@ -442,6 +496,14 @@ ul.social {
 	}
 }
 
+@media (max-width: 991px) {
```
Comment (jsvgoncalves) [inline source/css/custom.scss]: Is this related?

File: source/css/style.css [hunk 11]
```diff
@@ -243,6 +243,11 @@ h1, h2, h3, h4, h5, h6 {
 .paddings{
 	padding: 70px 0;
 }
+@media (max-width: 991px) {
```
Comment (jsvgoncalves) [inline source/css/style.css:246]: Why are we touching both `scss` and `css` files?

File: source/css/style.css [hunk 12]
```diff
@@ -1859,23 +1864,3 @@ aside .search .btn{
     color: #fff;
     font-size: 20pt;
 }
-
```
Comment (jsvgoncalves) [inline source/css/style.css:1862]: Why are we removing this?

File: source/github.php (hunk 2 above)
Comment (jsvgoncalves) [reply source/github.php]: Agreed

File: source/css/custom.scss [hunk 13]
```diff
@@ -36,14 +36,68 @@ a.strong {
...
+		color: #ee2d31;
+	}
+}
+
+.col-centered {
+	float: none;
+	margin: 0 auto;
+}
+
+/*	text-overflow: scroll;
```
Comment (jsvgoncalves) [inline source/css/custom.scss]: Remove

File: source/index.html.erb [hunk 14]
```diff
@@ -106,12 +106,22 @@ title: OpenShift Origin - Open Source Container Application Platform
   </div>
   <div class="">
     <div class="container wow fadeInUp">
-      <div class="row text-center">
-        <div class="col-xs-2 col-sm-2 col-md-2 col-lg-2"></div>
-        <div class="col-xs-8 col-sm-8 col-md-8 col-lg-8">
-          <pre class="code">$ oc cluster up    <a id="get-started" href="https://github.com/openshift/origin/blob/master/docs/cluster_up_down.md" target="_blank">Get Started</a></pre>
+      <div class="row">
```
Comment (jsvgoncalves) [inline source/index.html.erb:109]: What's happening here?

File: source/index.html.erb [hunk 15]
```diff
@@ -123,23 +133,21 @@ title: OpenShift Origin - Open Source Container Application Platform
...
-          <p>
-            Try out a fully functioning Origin instance with an integrated Docker registry, running locally on your machine with <%= link_to "Minishift", "/minishift/index.html" %>.
-          </p>
-        </div>
-        <div class="col-xs-12 col-sm-4 col-md-4 col-lg-4">
-          <h2>Download the Binary</h2>
-          <p>
-            Red Hat periodically publishes binaries to GitHub, which you can download on the OpenShift Origin <a href="https://github.com/openshift/origin/releases">Releases</a> page.
-          </p>
+        <div class="col-sm-10 col-centered">
```
Comment (jsvgoncalves) [inline source/index.html.erb:136]: What's happening here?

File: source/js/download.js [hunk 16]
```diff
@@ -0,0 +1,48 @@
+$(document).ready(function($) {
+    'use strict';
+
+    showdown.setFlavor('github');
+    var converter = new showdown.Converter({
+        headerLevelStart: 3
+    }),
+        php_url;
```
Comment (jsvgoncalves) [inline source/js/download.js]: `php_url = '/github.php';`

File: source/js/download.js [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+    // Included for local debugging, assuming running from Middleman, default port 4567
+    if (location.port == 4567) {
+        php_url = 'http://localhost:80/github.php';
+    } else {
```
Comment (jsvgoncalves) [inline source/js/download.js]: Remove the `else` block

File: source/js/download.js [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+        var jdata = JSON.parse(data),
+            linux_oc = $('#oc-platforms .linux-platform'),
+            mac_oc = $('#oc-platforms .mac-platform'),
+            windows_oc = $('#oc-platforms .windows-platform'),
+            linux_server = $('#server-platforms .linux-platform'),
+            asset, index, append_html;
+
+        console.log(jdata['assets']);
+        
+        for (var i = 0; i < jdata['assets'].length; i++) {
```
Comment (jsvgoncalves) [inline source/js/download.js]: What's happening here? Could use some comments

File: source/js/download.js [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+                append_html = '<a href="' + asset['browser_download_url'] + '">' + asset['name'].substring(index + 14) + '</a><br>';
+                if (asset['name'].indexOf('linux') >= 0) {
+                    linux_server.append(append_html);
+                }
+            }
+        }
+        var release_notes = $('#release-notes .release-content');
+        release_notes.html('<h2 style="text-align:center;">Release Notes</h2>' + converter.makeHtml(jdata['body']));
+    });
+});
```
Comment (jsvgoncalves) [inline source/js/download.js]: new line

File: source/layouts/layout.erb [hunk 20]
```diff
@@ -3,7 +3,7 @@
 <head>
   <meta charset="utf-8">
   <title><%= current_page.data.title || settings.site_title %></title>
-  <meta name="description" content="The next generation open source app hosting platform by Red Hat">
+  <meta name="description" content="<%= current_page.data.description || "The next generation open source app hosting platform by Red Hat" %>">
```
Comment (jsvgoncalves) [inline source/layouts/layout.erb:6]: relevant?

File: source/layouts/layout.erb [hunk 21]
```diff
@@ -100,6 +101,6 @@
     });
     </script>
     <!-- Place this tag right after the last button or just before your close body tag. -->
-<script async defer id="github-bjs" src="https://buttons.github.io/buttons.js"></script>
+  <script async defer id="github-bjs" src="https://buttons.github.io/buttons.js"></script>
   </body>
 </html>
```
Comment (jsvgoncalves) [inline source/layouts/layout.erb]: new line

**Review by jsvgoncalves — CHANGES_REQUESTED** [review]
Note for the future PRs
1. Generally it would be much easier to review if:
    1. There would be some comments on the more complicated portions of the code
    2. There would be a commit message, rather than just title, explaining in some detail what's happening. Most of the stuff is not obvious just by looking at the code.
3. Let's try to make the commit as atomic as possible, i.e., add only relevant code to the patch (`git add -p` might be useful)

In regards to the added `php` and `js` files, we should really be doing some error checking/handling, I have no idea what will happen if the requests will fail for some reason.

File: source/css/style.css (hunk 12 above)
Comment (wgordon17) [reply source/css/style.css:1862]: This was refactored into a reusable .code-button class

Comment (wgordon17) [reply source/css/style.css:1862]: Here, https://github.com/openshift/openshift.org/pull/68/commits/8507fb0e94ddf2ebdccf0014588caa8f49ed7d5d#diff-899fff31f3f49d91ccab380688c97248R60

File: source/index.html.erb (hunk 14 above)
Comment (wgordon17) [reply source/index.html.erb:109]: This was refactored from using placeholder columns to using a centered container column

File: source/index.html.erb (hunk 15 above)
Comment (wgordon17) [reply source/index.html.erb:136]: See here, https://github.com/openshift/openshift.org/pull/68/commits/8507fb0e94ddf2ebdccf0014588caa8f49ed7d5d#r107411653

File: source/layouts/layout.erb (hunk 20 above)
Comment (wgordon17) [reply source/layouts/layout.erb:6]: For SEO purposes, and search engine results, I would think that an appropriate description be available on the downloads page, different from the description on the index page

File: source/index.html.erb (hunk 14 above)
Comment (jsvgoncalves) [reply source/index.html.erb:109]: I'm fine with this change, but as discussed before, let's try to have logically separate commits (and PRs)

File: source/css/style.css (hunk 11 above)
Comment (wgordon17) [reply source/css/style.css:246]: The changes were made in whichever existing file the class was in, with new additions being opted for the scss file. The .paddings class already existed in the .css file, so that was where I left the style. I'm unsure of why this was originally setup with css and scss files

File: README.md (hunk 1 above)
Comment (wgordon17) [reply README.md]: Wasn't aware of the option! I've refactored to use middleman-php...thanks for this!

File: source/layouts/layout.erb (hunk 20 above)
Comment (jsvgoncalves) [reply source/layouts/layout.erb:6]: >  I'm fine with this change, but as discussed before, let's try to have logically separate commits (and PRs)

File: source/css/style.css (hunk 11 above)
Comment (jsvgoncalves) [reply source/css/style.css:246]: I guess the problem was that we incorrectly started editing the `css` file at some point in the past, so now it's hard to incorporate the changes backwards. We should probably clean up the whole thing at some point.

**Review by jsvgoncalves — APPROVED** [review]

File: source/js/download.js [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+            var downloads = $('#downloads'),
+                more_details = $('#more-details');
+            
+            downloads.html('<h2>Sorry, we appear to be having difficulties fetching our binaries</h2>' + more_details[0].outerHTML);
+        };
+
+    $.ajax({
+        url:'/github.php',
+        success: function(data) {
+            var jdata = JSON.parse(data),
```
Comment (jsvgoncalves) [inline source/js/download.js]: There's still the issue here, what if the response cannot be parsed as JSON? I've tested and the page does show, but with no download links.

**Review by jsvgoncalves — CHANGES_REQUESTED** [review]

**Review by jiri-fiala — APPROVED** [review]

File: .gitignore [hunk 23]
```diff
@@ -16,6 +16,7 @@
 # Ignore cache
 /.sass-cache
 /.cache
+/source/github_api_output.json
```
Comment (jsvgoncalves) [inline .gitignore]: The file is generated on root level, so this needs to be `github_api_output.json` (without the `source/` dir)

**Review by jsvgoncalves — CHANGES_REQUESTED** [review]

File: source/js/download.js (hunk 22 above)
Comment (jsvgoncalves) [reply source/js/download.js]:
```js
success: function(data) {
        try {
            var jdata = JSON.parse(data);
        } catch(err) {
            display_error_msg();
            return;
        }
        var linux_oc = $('#oc-platforms .linux-platform'),
        ///...
```

It might also make sense to do something like: 

```js
success: function(data) {
    try {
        parse_data(data);
    } catch(err) {
        display_error_msg();
    }
},
```

And have the `parse_data()` function raising any errors it finds.

File: source/js/download.js [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+                    } else if ((index = asset['name'].indexOf('origin-server')) >= 0) {
+                        // index + 14 encapsulates the "origin-server" length plus an additional character for the trailing "-"
+                        append_html = '<a href="' + asset['browser_download_url'] + '">' + asset['name'].substring(index + 14) + '</a><br>';
+                        if (asset['name'].indexOf('linux') >= 0) {
+                            linux_server.append(append_html);
+                        }
+                    }
+                }
+            } else {
+                display_error_msg();
```
Comment (jsvgoncalves) [inline source/js/download.js:55]:
```js
} else {
    display_error_msg();
    return;
}
```

**Review by jsvgoncalves — CHANGES_REQUESTED** [review]

File: source/js/download.js (hunk 24 above)
Comment (wgordon17) [reply source/js/download.js:55]: Good catch, thanks!

**Review by jsvgoncalves — APPROVED** [review]

**Comment by jsvgoncalves** [comment]
@wgordon17 LGTM, please squash the commits so that we can merge the PR.
